Japanese search feature

akr4 opened the pull request:
This PR adds a Japanese text search feature, which uses `text_ja` type in Solr and provides better results.

# Changes

## Library dependency
- Added [language-detector](https://github.com/optimaize/language-detector)

## Solr schema
- Added `text_ja` type
- Added `language`, `content_ja` fields
- Updated schema version from 2.1 to 2.2.

## Index and query
The new process of index and query:
1. `Ingester` detects languages
2. `Ingester` indexes chunks and mini-chunks
  - If the language is not Japanese, indexing is done in the same way as the current released version
  - If the language is Japanese
    - Do the same things as for non-Japanese documents
    - Store the content and the filename in the `content_ja` field
    - Store the language (`"ja"`) in the `language` field
    - Index mini-chunks

3. `LuceneQuery` and `HighlightedText` uses `text` and `content_ja` for query and highlighting.

## Mini-chunk approach

Mini-chunks are placed at the windows of chunks. Figure 1 shows that.
![fig1](https://user-images.githubusercontent.com/10892/61676864-6e312880-ad38-11e9-8db0-b07b3c9b4b46.png)

In Figure 2, a hit is in the base part of chunk <span>#</span>2 (not window). This is the only chunk which has hits, so we should pick chunk <span>#</span>2.
![fig2](https://user-images.githubusercontent.com/10892/61676869-76896380-ad38-11e9-9b5d-939dd72ebaeb.png)

Figure 3 is a particular version of Figure 2. The hit is on the border of the base and the window of chunk <span>#</span>2. As of Figure 2, we should pick chunk <span>#</span>2.
![fig3](https://user-images.githubusercontent.com/10892/61676875-7d17db00-ad38-11e9-95dc-d556e822dca3.png)

In Figure 4, a hit is in the base part of chunk <span>#</span>3, but in the window of chunk <span>#</span>2 at the same time. Both chunk <span>#</span>2 and <span>#</span>3 have hits. In this case, we can know the hit is in the window of chunk <span>#</span>2 because of mini-chunk <span>#</span>2 also have a hit. Hits in mini-chunks means hits are in windows. So we can pick chunk <span>#</span>3 only.
![fig4](https://user-images.githubusercontent.com/10892/61676884-886b0680-ad38-11e9-8d6b-048c90611bf9.png)

In Figure 5, chunk <span>#</span>3 has three hits, and mini-chunk <span>#</span>2 has two hits. We can calculate how many hits exist in chunk <span>#</span>2 by the following formula:
```
<hit count in the base chunk> = <hit count in the chunk> - <hit count in the mini-chunk>
```
In this case, it turns out that the base part of chunk <span>#</span>2 has 1 (= 3 - 2) hit.
So we can pick chunk <span>#</span>2, and chunk <span>#</span>3.
![fig5](https://user-images.githubusercontent.com/10892/61676899-8f921480-ad38-11e9-9b1e-74ac37fde027.png)

## Termfreq
To determine whether or not chunks should be picked up in a search result, This code uses `termfreq` function in Solr query.
```
sum(
  termfreq("content_ja","term1"),
  termfreq("content_ja","term2")
)
```
`term1` and `term2` are terms included in a query. For example, 雨が降っ ている (It's raining) is divided into two terms 雨 and 降る. If your query has some terms, `termfreq` should be used for each term. So knowing these terms before the query is required. That is achieved by making an additional request to Solr. `QueryParse` is the class for that.

esaunders commented on KeywordSearch/src/org/sleuthkit/autopsy/keywordsearch/HighlightedText.java:
```diff
@@ -359,16 +362,33 @@ public String getText() {
-                q.addFilterQuery(filterQuery);
-                q.addHighlightField(highlightField);
-                q.setHighlightFragsize(0); // don't fragment the highlight, works with original highlighter, or needs "single" list builder with FVH
+                if (2.2 <= indexSchemaVersion) {
```
I see references to schema version 2.2 but the schema.xml file has not been updated.

akr4 replied:
Thank you for your review.
I updated `KeywordSearch/release/solr/solr/configsets/AutopsyConfig/conf/schema.xml`. Isn't that sufficient?

esaunders replied:
Yes, but I don't see the change in schema.xml above.

akr4 replied:
The path in my previous comment is wrong. The correct path is `KeywordSearch/solr/solr/configsets/AutopsyConfig/conf/schema.xml
`. 
Diff:
https://github.com/sleuthkit/autopsy/pull/5139/files#diff-3f316c678543ca04318ac83da0a094d8

esaunders replied:
When I click on the link you provided (on this side of the world :-) I see the changes to add the text_ja type and the content_ja and language fields but I do not see any update to the schema version. When I look at the entire file containing your changes I see the following:

<schema name="Autopsy Keyword Search" version="2.1">

What do you see?

akr4 replied:
I noticed that Autopsy is using the version attribute of `<schema>` tag. I will update it later. Thank you for pointing it out.

esaunders commented:
Highlighting of English text does not appear to work correctly. I added the attached eng_utf16l.txt file as a Logical File and ran an exact match search for the string "self-treatment". Autopsy 4.12 correctly highlights 2 hits in the same document whereas the code on this branch results in 7 hits (see attachments).
![autopsy-4-12-eng-search-highlighting](https://user-images.githubusercontent.com/1292764/64553788-0d6da780-d308-11e9-8fb5-4b39c36d603e.PNG)
[eng_utf16l.txt](https://github.com/sleuthkit/autopsy/files/3591861/eng_utf16l.txt)

esaunders commented:
New source files are missing copyright/license header block.

esaunders commented:
When I index the attached file and search for 東京都 Autopsy reports 2 matches but I only see highlighting for a single match.
![jp-wikipedia-search-highlighting](https://user-images.githubusercontent.com/1292764/64558382-b79dfd00-d311-11e9-9c32-d75a660414ef.PNG)
[Japanese.Wikipedia.utf8.txt](https://github.com/sleuthkit/autopsy/files/3592112/Japanese.Wikipedia.utf8.txt)

esaunders commented on KeywordSearch/src/org/sleuthkit/autopsy/keywordsearch/Server.java:
```diff
@@ -175,6 +188,12 @@ public String toString() {
                 return "chunk_size"; //NON-NLS
             }
+        },
+        TERMFREQ {
```
This doesn't appear to be a field in the schema.

akr4 replied:
`termfreq` is a function to get the number of occurrence of the term. It's not an actual field in indices, but it can be gotten from documents in the same way as fields.
```document.getFieldValue(Server.Schema.TERMFREQ.toString())```
Should `Server.Schema` have actual fields only?

esaunders replied:
If it's coming back as a field in the document I'm fine with leaving it where it is. I would suggest that you add a comment to avoid confusion in the future if someone is looking at the code and wondering why there is no field with that name in the schema :-)

akr4 replied:
I added a comment on `TERMFREQ`. Thank you for your suggestion.

akr4 commented:
@esaunders Thank you for your review. I will investigate the detail of "self-treatment" and 東京都 issues, and get back to you.

esaunders commented:
I started my testing using a small data set that I have shared with you through another channel and I noticed the following...

The Chinese documents are being identified as Japanese (perhaps not so surprising) and the Japanese documents are not being identified as Japanese (perhaps the word "NOVA" at the beginning of the document is throwing language detection off or maybe there isn't enough content for the language detector to work with).

I understand that language detection can be difficult (especially with a small data set) but it might be worthwhile taking a look into this to see if there is anything we can do about it.

esaunders commented on KeywordSearch/src/org/sleuthkit/autopsy/keywordsearch/LanguageDetector.java:
```diff
@@ -0,0 +1,55 @@
+
+class LanguageDetector {
+
+  public enum Language {
```
I find the existence of 2 different enums named "Language" to be confusing. Is it possible to consolidate these enums and possible use Enum.name() and toString() methods in place of static strings (e.g. "en" and "ja" below)?

akr4 replied:
I consolidated the two `Language` classes. Now there is only `JAPANESE`.  A `Language` is detected means the document should be indexed to a language-specific field such as `content_ja`.

esaunders commented on KeywordSearch/src/org/sleuthkit/autopsy/keywordsearch/LanguageDetector.java:
```diff
@@ -0,0 +1,55 @@
+    return languageDetector.detect(textObject).transform(Optional::of).or(Optional.empty()).map(LdLocale::getLanguage).flatMap(LanguageDetector::toLanguage);
+  }
+
+  private static Optional<Language> toLanguage(String s) {
```
Perhaps this can be replaced by something else in the consolidated enum I suggest above. There are a number of different ways you could do this, e.g. you might leverage Enum.valueOf().

esaunders commented on KeywordSearch/src/org/sleuthkit/autopsy/keywordsearch/QueryParser.java:
```diff
@@ -0,0 +1,63 @@
+   */
+  static Result parse(String query, List<Server.Schema> fields) throws KeywordSearchModuleException, NoOpenCoreException {
+    SolrQuery q = new SolrQuery();
+    q.setShowDebugInfo(true);
```
Does debug info really need to be turned on here?

akr4 replied:
Yes. I use the debug flag to get parsed query information here.
https://lucene.apache.org/solr/7_5_0//solr-core/org/apache/solr/handler/FieldAnalysisRequestHandler.html
But there is the non-debug API `FieldAnalysisRequestHandler` for the same purpose. I will come up with a commit which uses it.

akr4 replied:
Changed to use `FieldAnalysisRequestHandler`. The result is the same but it's a official API (not debug).

esaunders commented on KeywordSearch/src/org/sleuthkit/autopsy/keywordsearch/HighlightedText.java:
```diff
@@ -359,16 +362,33 @@ public String getText() {
+                    q.addField(Server.Schema.LANGUAGE.toString());
+                    // in case of single term literal query there is only 1 term
+                    LanguageSpecificContentQueryHelper.configureTermfreqQuery(q, keywords.iterator().next());
+                    q.addFilterQuery(filterQuery);
```
These 2 lines appear to be the same in both the "if" and "else" blocks. Perhaps they should be moved out of the if..else.

esaunders commented on KeywordSearch/src/org/sleuthkit/autopsy/keywordsearch/HighlightedText.java:
```diff
@@ -406,12 +426,40 @@ public String getText() {
-                    List<String> contentHighlights = responseHighlightID.get(LuceneQuery.HIGHLIGHT_FIELD);
-                    if (contentHighlights == null) {
-                        highlightedContent = attemptManualHighlighting(response.getResults(), highlightField, keywords);
+                    SolrDocument document = response.getResults().get(0);
```
It might make it a little easier to read this method if this block of code related to mini chunk highlighting was factored out into its own method.

esaunders commented on KeywordSearch/src/org/sleuthkit/autopsy/keywordsearch/HighlightedText.java:
```diff
@@ -406,12 +426,40 @@ public String getText() {
-                        highlightedContent = attemptManualHighlighting(response.getResults(), highlightField, keywords);
+                    SolrDocument document = response.getResults().get(0);
+                    Object language = document.getFieldValue(Server.Schema.LANGUAGE.toString());
+                    if (2.2 <= indexSchemaVersion && language != null) {
```
If language != null do we really need the schema version check?

akr4 commented:
About `self-treatment` issue, I had changed not using phrase queries by mistake. As a result of that, `self-treatment` matches `self` and `treatment` respectively. I fixed this problem in my environment. I will push once finishing the 東京都 issue.

akr4 commented:
I pushed new commits. This inclue:
- Enabled phrase query ("self-treatment" query now works fine)
- Use `original` highlighter for Japanese text
- Updated the version of `<schema>` tag.
- Added copyright/license block
- Added a comment on `TERMFREQ`
- Consolidated the Language classes

Now "self-treatment" issue is fixed by using phrase queries. However, using phrase queries on Japanese texts caused another problem. It seems that `FastVectorHighlighter` is not work in some cases. So I changed the code to use `original` highlighter when `language` value is `ja`.

Even with these changes, 東京都 issue still exists.
東京都 is divided into 東京 and 都 before generating snippets and matches individually. It looks like a single match, but two matches 東京 and 都 happen consecutively. 
New `unified` highlighter (from Solr 6.4) handles this case as expected. Once we migrate to Solr 8, I think it's good to consider to use it.
So for now, as a workaround, it would be possible to change the style of the highlighted words to show there're multiple matches.

esaunders commented:
Thanks for the update. At this point I'm waiting until this pull request is as close as possible to a "done" state before I take a look at it again. Let me know when you get to a point where you believe all of feedback has been addressed (including Codacy issues).

akr4 commented:
I pushed some commits:
- I updated `LanguageDetector` to use all language profiles which the implementation provides. I saw it now can detect Chinese text correctly in my environment.
- Use `FieldAnalysisRequestHandler` to parse queries instead of the debug info.

I will work on the Codacy result later.

akr4 commented:
@esaunders Hi. Now I finished all the things to do.
appveyor and travis-ci still failed, but it seems not to be related to these changes.

akr4 commented:
@esaunders Thank you for your review. What is the next step for this PR to be merged?

rcordovano commented:
@Akr4, @esaunders, and @bcarrier, I will merge this into the develop branch now. It will not be run through automated regression testing until next week, since we are currently using the testing infrastructure to validate the Autopsy 4.13.0 release.

akr4 commented:
@rcordovano Thank you for the information.

rcordovano commented:
You are welcome, @akr4.
